midend/replaceSelectRange.cpp

```diff
@@ -0,0 +1,177 @@
+/*
+* Copyright 2020, MNK Labs & Consulting
+* http://mnkcg.com
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*
+*/
+
+#include "replaceSelectRange.h"
+
+namespace P4 {
+
+static int trailingZeros(big_int n, int width) {
+    int zeros = 0;
+
+    while (n > 0 && (n & 1) == 0) {
+        zeros += 1;
+        n >>= 1;
+    }
+
+    return (zeros < width) ? zeros : width;
+}
+
+bool DoReplaceSelectRange::checkRange(const IR::Range* range) {
+    if (range == nullptr)
+        return false;
+
+    if (!range->left->is<IR::Constant>()) {
+        std::cout << "must evaluate to a compile-time "
+                  << "constant" << range->left << std::endl;
+        return false;
+    }
+    auto left = range->left->to<IR::Constant>()->value;
+    if (!range->right->is<IR::Constant>()) {
+        std::cout << "must evaluate to a compile-time "
+                  << "constant" << range->right <<  std::endl;
+        return false;
+    }
+    auto right = range->right->to<IR::Constant>()->value;
+    if (right < left) {
+        std::cout << "Range end " << std::hex << std::showbase << right
+        << " is less than start " << left << std::endl;
+        return false;
+    }
+    return true;
+}
+
+std::vector<const IR::Mask *>
+DoReplaceSelectRange::rangeToMasks(const IR::Range *r) {
+    bool st = checkRange(r);
+    BUG_CHECK(st, "Range check failed");
```

Comment:
This cannot be a bug, people can write programs with such ranges and the compiler should not crash.